Allow search page widget search field hint text and keyboard text input action to be customizable

samlythemanly opened the pull request:
## Description

Currently, the search page widget's hint text and keyboard text input action are hardcoded to the material localization of 'Search' and the search text input action, respectively. This prevents further customization where search functionality is desired, but the nuance of the action isn't exactly "searching" (e.g. adding a location via Google Maps Places API autocomplete).

Added a `getHintText` function (that takes in a `BuildContext`) and a `textInputAction` field to `SearchDelegate`. These default to the previous hardcoded attributes that were present in `SearchPage`.

These changes are purely aesthetic changes; no additional functionality is provided.
All tests still pass.

## Related Issues

N/A

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

justinmc commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -89,6 +89,19 @@ Future<T> showSearch<T>({
+  /// enter an input.
+  ///
+  /// Defaults to [MaterialLocalizations.searchFieldLabel].
+  String getHintText(BuildContext context) =>
```
What was your motivation for making this a function?  The alternative would be to pass it in as a String and use the localization value if it's null.  Same thing either way but I'm wondering if you have any specific use for it being a function.

somarkoe replied:
Either works, but I do agree, your alternative is much cleaner. Fixing!

justinmc reviewed (commented):
@somarkoe Thanks for submitting this and sorry for the lack of reviewers.  Looks like a simple improvement to make Search a lot more useful, I'm all on board.

Just a discussion below to decide the best way to do this.  There was a Linux test failing, but I think it was flaky so I reran it.  We'll see.

MichaelPriebe commented:
While you're at it, please do keyboardType: https://github.com/flutter/flutter/issues/29983
You don't always search with text.

justinmc commented:
I agree with the idea to add keyboardType to this as well.

Looks like all tests are passing now.  @somarkoe are you still available to get this PR in?

somarkoe commented:
Yes! Sorry, been swamped by work and haven't touched the project that deals with this in a while. I'll add the keyboard type request to it as well.

justinmc commented:
Awesome, thank you @somarkoe.  Update the PR when you get a chance and I'll be happy to help you get this in.

MichaelPriebe commented:
@justinmc Perhaps will you be able to get https://github.com/flutter/flutter/pull/28165 in too? There's so much about the search page that is hard coded and it shouldn't be.

justinmc commented:
Yeah that one looks straightforward to me.  The original reviewer @HansMuller isn't around today but I'll talk it through with him when he's back.

MichaelPriebe commented:
Abandoned? :(

samlythemanly committed:
Allow search page widget search field hint text, keyboard text input action, and keyboard text input type to be customizable.

[Description]
Currently, the search page widget's hint text and keyboard text input action are hardcoded to the material localization of 'Search' and the search text input action, respectively, and the text input type defaults to that defined in `TextField`. This prevents further customization where search functionality is desired, but the nuance of the action isn't exactly "searching" (e.g. adding a location via Google Maps Places API autocomplete).

[Changes]
Added `hintText`, `textInputAction`, and `keyboardType` fields to `SearchDelegate`. These default to the previous hardcoded attributes that were present in `SearchPage` (or `null`).

[Testing]
Added a new unit test to cover the new `textInputAction` case (and modified the existing one's name to indicate it's the default behavior).

All existing tests still pass, as well.

somarkoe commented:
Oops, been a while since I've used git.

Reopened and made the relevant changes. Sorry it took so long!

MichaelPriebe commented:
😍 merge this ish!

MichaelPriebe commented:
Only thing that could be added on is https://github.com/flutter/flutter/pull/30388 / https://github.com/flutter/flutter/pull/28165 (seems like they are the same 1 line of code)

justinmc commented:
@somarkoe Thanks for following up and no worries about the delay!

Sorry to drag this out longer, but I realized that this should update the tests as well.  [This test](https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/search_test.dart#L468) should probably say "keyboard shows search button by default" now.

Then, could you add a separate test that tests a different `textInputAction`?  Just duplicate that other test and add `textInputAction` to the `TestHomePage` parameters.

After that this is good to go and we can merge.

justinmc commented:
@MichaelPriebe I'll try to see if we can get things moving on those 2 other PRs you referenced.

MichaelPriebe commented:
@justinmc how's this going?

justinmc commented:
Waiting on @somarkoe for the changes mentioned above.

MichaelPriebe commented:
@somarkoe come thru

somarkoe commented:
Oops, didn't notice the warning not to override, but also unsure how to alleviate this problem in the test. Thoughts?

MichaelPriebe commented:
@justinmc What do you think?

somarkoe commented:
I made the fields non-final since apparently overriding fields when a class is extended is no bueno, and this abstract class is one meant to be extended and not implemented (since it has functions with default behaviors).

MichaelPriebe commented:
Excited for this, can finally ditch my very hacky extension of the search delegate.

MichaelPriebe commented:
@justinmc Is it ready now?

MichaelPriebe commented:
Lets get this pull a'mergin'!

justinmc commented on packages/flutter/lib/src/material/search.dart:
```diff
@@ -89,6 +89,23 @@ Future<T> showSearch<T>({
 /// for another [showSearch] call.
 abstract class SearchDelegate<T> {
 
+  /// The type of action button to use for the keyboard.
```
We could keep the fields `final` by using a constructor instead like this:

```dart
  const SearchDelegate({
    this.keyboardType,
    this.hintText,
    this.textInputAction = TextInputAction.search,
  });
```

Then below, make all of the fields `final` and remove the default assignment from `textInputAction`.

justinmc commented on packages/flutter/test/material/search_test.dart:
```diff
@@ -639,7 +653,12 @@ class _TestSearchDelegate extends SearchDelegate<String> {
+    if (textInputAction != null) {
+      this.textInputAction = textInputAction;
+    }
+  }
```
Then with the change in my other comment, you can user `super` here:

```dart
...
}) : super(
  textInputAction: textInputAction,
);

justinmc reviewed (commented):
Sorry for the delay!  I talked to @HansMuller to get some tips for the non-final problem.

MichaelPriebe commented:
@somarkoe What do you think about the suggestions?

MichaelPriebe commented:
@somarkoe Would love to hear what you think about the suggested changes.

somarkoe commented:
So sorry! Was on vacation and now I'm moving 😅 

Ill take a look ASAP!

shihaohong commented:
I'm going to close this since https://github.com/flutter/flutter/pull/36409 has been merged. Thank you @somarkoe for your initial work with this!
